sdk/core/azure-core/azure/core/pipeline/policies/authentication.py

```diff
@@ -60,6 +68,7 @@ def send(self, request):
         :return: The pipeline response object
         :rtype: ~azure.core.pipeline.PipelineResponse
         """
-        token = self._credential.get_token(*self._scopes)
-        self._update_headers(request.http_request.headers, token)  # type: ignore
+        if self._cached_token_expired:
+            self._token = self._credential.get_token(*self._scopes)
```

Comment:
Can we move the if self._cached_token_expired check into get_token method and get_token always returns a non-expiry token?

Reply:
> Can we move the if self._cached_token_expired check into get_token method 

It's already there--credentials have their own caches. Those are more expensive because they support more complex scenarios, e.g. arbitrary scopes. Scopes are baked into this policy however, so its arguments to `get_token` will never change, and it now knows how long tokens are valid. So it may as well hang on to the last token it received and reuse it rather than ask the credential for a token every time a request passes through.

> and get_token always returns a non-expiry token?

I don't follow--it does now, unless AAD returns an expired token?

Reply:
I mean 
we have the check in
self._token = self._credential.get_token(*self._scopes)
which guarantees a non-expiry token.
Why we use 
if self._cached_token_expired:
            self._token = self._credential.get_token(*self._scopes)
rather than 
self._token = self._credential.get_token(*self._scopes) 
only?

Reply:
The issue isn't whether `get_token` might return an expired token. It's that we execute this code on every request, and calling `get_token` is more expensive than checking the time.

Reply:
Not sure if i understand correctly, get_token also has a cache and checks cache before calling server to get a token, right?

Reply:
That's correct.

Reply:
So it looks to me our logic is
check cache
if fail -> call get_token()
in get_token check cache
if fail -> call server

I think we can call get_token() w/o the first checking cache?
The benefit is
1. Reduce maintenance cost if we change the design in the future.
2. Make the code clean and have one place to handle token
3. Performance (can ignore. :))

I will leave you to make to call.

Reply:
> So it looks to me our logic is
> check cache
> if fail -> call get_token()
> in get_token check cache
> if fail -> call server

It's a bit misleading to describe the policy as having a cache (so I renamed the check). It's just the last token received.

I benchmarked the best case cache lookup in `get_token` (one cached token and it's a hit). It's about 9x slower on my machine than this check in the bearer policy. It's a matter of microseconds but I'll take it.

I don't think the code is complex. `AccessToken` was added to support other scenarios, not to speed up `BearerTokenCredentialPolicy`. So I'm adding a couple lines of code to the policy to take advantage of available metadata, not adding a new class to expose metadata so I can realize a small performance gain in the policy.